lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Task/TaskStatus.cs

```diff
@@ -98,28 +98,48 @@ public void SetException(Exception e)
         {
             lock (_heartBeatManager)
             {
-                try
+                if (_lastException.IsPresent())
                 {
-                    if (!_lastException.IsPresent())
-                    {
-                        _lastException = Optional<Exception>.Of(e);
-                    }
+                    return;
+                }
 
-                    State = TaskState.Failed;
+                _lastException = Optional<Exception>.Of(e);
+                State = TaskState.Failed;
+
+                try
+                {
                     _taskLifeCycle.Stop();
                 }
-                finally
+                catch (Exception stopException)
                 {
-                    Heartbeat();
+                    Utilities.Diagnostics.Exceptions.Caught(
```

Comment:
@markusweimer Sorry, `TaskStopHandlers` is actually a Set. The default `TaskStopHandlers` is an empty set, in which it doesn't do anything. `TaskClose` and `TaskStop` are different. `TaskClose` is triggered by the user's Driver while `TaskStop` is always called on Task end. This makes sense for the `DefaultCloseHandler` to kill the Evaluator, since that means receiving a `CloseEvent` on the Task side was not expected; However, `TaskStop` is always expected so the default action should be to do nothing.